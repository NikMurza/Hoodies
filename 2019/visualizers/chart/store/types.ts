import { ChartData } from '@common/api/types/chart';

export interface ChartVisualizerInstance {
  openId: string;
  data?: ChartData;
  isFetching: boolean;
}

export interface ChartVisualizersState {
  [instanceId: string]: ChartVisualizerInstance;
}

export interface StatePart {
  chartVisualizers?: ChartVisualizersState;
}
export type GetState = () => StatePart;

export interface ChartPayload<T> {
  instanceId: string;
  data?: T;
}
