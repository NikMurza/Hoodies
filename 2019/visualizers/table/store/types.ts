import { PPSom } from '@common/api/types/PPSom';

export interface TableVisualizerInstance {
  openId: string;
  tabSheetData: TabSheetRawData;
  isFetching: boolean;
}

export interface TableVisualizersState {
  [instanceId: string]: TableVisualizerInstance;
}

export interface StatePart {
  tableVisualizers?: TableVisualizersState;
}
export type GetState = () => StatePart;

export interface TabSheetPayload<T> {
  instanceId: string;
  data?: T;
}

export interface TabSheetRawData {
  /** Структура и диапазоны данных. */
  Data?: PPSom.TabData[];
  /** Загруженные диапазоны. */
  Ranges?: PPSom.TabRange[];
  /** Метаданные */
  Metadata?: PPSom.TabMd;
  /** Выделение (Дублируется из метаданных). */
  Selection?: PPSom.TabSelection;
}
