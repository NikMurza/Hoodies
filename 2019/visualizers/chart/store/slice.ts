import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { ChartData } from '@common/api/types/chart';
import { ChartVisualizerInstance, ChartVisualizersState, StatePart, ChartPayload } from './types';

// общий стейт всех инстансов - просто мапа
const initialState: ChartVisualizersState = {};

// стейт конкретного инстанса
export const initialInstanceState: Readonly<ChartVisualizerInstance> = {
  openId: '',
  data: {
    json: null,
    image: '',
    toolTipMap: '',
  },
  isFetching: false,
};

export const { actions, reducer } = createSlice({
  name: 'chartVisualizers',
  initialState,
  reducers: {
    createInstance(state: ChartVisualizersState, { payload }: PayloadAction<string>): void {
      state[payload] = { ...initialInstanceState };
    },
    setOpenId(
      state: ChartVisualizersState,
      { payload }: PayloadAction<{ instanceId: string; openedId: string }>,
    ): void {
      state[payload.instanceId].openId = payload.openedId;
    },
    setChartData(
      state: ChartVisualizersState,
      { payload }: PayloadAction<ChartPayload<ChartData>>,
    ): void {
      state[payload.instanceId].data = payload.data;
    },
    setIsFetching(
      state: ChartVisualizersState,
      { payload }: PayloadAction<{ instanceId: string; isFetching: boolean }>,
    ): void {
      state[payload.instanceId].isFetching = payload.isFetching;
    },
  },
});

// selectors
const instance = (state: StatePart, instanceId: string): ChartVisualizerInstance =>
  state.chartVisualizers[instanceId] || null;

const openId = (state: StatePart, instanceId: string): string =>
  state.chartVisualizers[instanceId]?.openId;

const chartData = (state: StatePart, instanceId: string): ChartData =>
  state.chartVisualizers[instanceId]?.data;

const isFetching = (state: StatePart, instanceId: string): boolean =>
  state.chartVisualizers[instanceId]?.isFetching;

export const selectors = {
  instance,
  openId,
  chartData,
  isFetching,
};

const spinner = {
  selectors,
  actions,
  reducer,
};
export default spinner;
