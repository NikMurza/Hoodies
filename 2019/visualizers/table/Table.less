@import (reference) '~@common/styles/constants/colors.less';
@import (reference) '~@common/styles/constants/margins.less';

.table-visualizer {
  position: absolute;
  width: 640px - @container-border-width * 2;
  height: 560px - @container-border-width * 2;

  .table-visualizer-tabsheet {
    .PPTSContainer {
      table.PPTSGrid > tbody > tr:first-child {
        display: none;
      }
    }
  }
}

.with-move,
.with-resize,
.visualizer-wrapper {
  .table-visualizer {
    position: static;
  }
}
