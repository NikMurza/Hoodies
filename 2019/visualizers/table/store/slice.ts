import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import {
  TableVisualizersState,
  TableVisualizerInstance,
  TabSheetPayload,
  TabSheetRawData,
  StatePart,
} from './types';

// общий стейт всех инстансов - просто мапа
const initialState: TableVisualizersState = {};

// стейт конкретного инстанса
export const initialInstanceState: Readonly<TableVisualizerInstance> = {
  openId: '',
  tabSheetData: {},
  isFetching: false,
};

export const { actions, reducer } = createSlice({
  name: 'tableVisualizers',
  initialState,
  reducers: {
    createInstance(state: TableVisualizersState, { payload }: PayloadAction<string>): void {
      state[payload] = { ...initialInstanceState };
    },
    setOpenId(
      state: TableVisualizersState,
      { payload }: PayloadAction<{ instanceId: string; openedId: string }>,
    ): void {
      state[payload.instanceId].openId = payload.openedId;
    },
    setTabSheetData(
      state: TableVisualizersState,
      { payload }: PayloadAction<TabSheetPayload<TabSheetRawData>>,
    ): void {
      state[payload.instanceId].tabSheetData = payload.data;
    },
    setIsFetching(
      state: TableVisualizersState,
      { payload }: PayloadAction<{ instanceId: string; isFetching: boolean }>,
    ): void {
      state[payload.instanceId].isFetching = payload.isFetching;
    },
  },
});

// selectors
const instance = (state: StatePart, instanceId: string): TableVisualizerInstance =>
  state.tableVisualizers[instanceId] || null;

const openId = (state: StatePart, instanceId: string): string =>
  state.tableVisualizers[instanceId]?.openId;

const tabSheetData = (state: StatePart, instanceId: string): TabSheetRawData =>
  state.tableVisualizers[instanceId]?.tabSheetData;

const isFetching = (state: StatePart, instanceId: string): boolean =>
  state.tableVisualizers[instanceId]?.isFetching;

export const selectors = {
  instance,
  openId,
  tabSheetData,
  isFetching,
};

const spinner = {
  selectors,
  actions,
  reducer,
};
export default spinner;
