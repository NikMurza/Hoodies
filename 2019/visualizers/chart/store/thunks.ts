import { IAsyncHandler } from '@common/containers/spinner/store/thunks';
import api from '@common/api/bi/visualizers/root';
import { GetState } from './types';
import slice from './slice';

export const load = (instanceId: string): IAsyncHandler => async (
  dispatch: Function,
  getState: GetState,
): Promise<void> => {
  try {
    const state = getState();
    dispatch(slice.actions.setIsFetching({ instanceId, isFetching: true }));
    const openId = slice.selectors.openId(state, instanceId);

    const data = await api.getChart(openId, instanceId);

    dispatch(slice.actions.setChartData({ instanceId, data }));
  } finally {
    dispatch(slice.actions.setIsFetching({ instanceId, isFetching: false }));
  }
};

export default {
  load,
};
