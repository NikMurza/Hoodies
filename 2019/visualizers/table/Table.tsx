import * as React from 'react';
import TabSheet from '@common/components/lib/rich/tab-sheet/TabSheet';
import { connect } from 'react-redux';
import { compose } from '@reduxjs/toolkit';
import { ETabSheetEditMode } from '@common/components/lib/enums/tab-sheet';
import { PPSom } from '@common/api/types/PPSom';
import { ITabSheetPictureMouseDownArgs } from '@common/components/lib/types/tab-sheet';
import WithFetching from '@common/components/lib/canvas/components/with-fetching/WithFetching';
import { VisualizerType } from '@common/api/types/visualizers/root';
import tableVisualizerSlice from './store/slice';
import tableVisualizerThunk from './store/thunks';
import { TabSheetRawData, StatePart } from './store/types';
import './Table.less';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const Component = WithFetching<any>(TabSheet);

interface OwnProps {
  openedId: string;
  instanceId: string;
  // при изменении этого флага - будет вызываться load
  triggerReload?: boolean;
  browserRoutingMode?: boolean;
  x?: number;
  y?: number;
  width?: number;
  height?: number;
  zoom?: number;
}

interface StateProps {
  instanceOpenedId: string;
  tabSheetData: (instanceId: string) => TabSheetRawData;
  isFetching: boolean;
}

interface DispatchProps {
  setOpenId: (payload: { instanceId: string; openedId: string }) => void;
  load: (instanceId: string) => void;
  createInstance: (instanceId: string) => void;
  pressTabsheetPicture: (instanceId: string, args: ITabSheetPictureMouseDownArgs) => void;
  loadTabSheetRanges: (instanceId: string, ranges?: PPSom.Rect[]) => void;
}

type Props = OwnProps & StateProps & DispatchProps;

export const Table: React.FC<Props> = (props: Props): JSX.Element => {
  const {
    x,
    y,
    width,
    height,
    zoom,
    browserRoutingMode,
    instanceId,
    openedId,
    instanceOpenedId,
    triggerReload,
    isFetching,
    tabSheetData,
    createInstance,
    setOpenId,
    load,
    pressTabsheetPicture,
    loadTabSheetRanges,
  } = props;

  React.useEffect((): void => {
    if (instanceId && openedId) {
      createInstance(instanceId);
      setOpenId({ instanceId, openedId });
    }
  }, [instanceId, openedId]);

  React.useEffect((): void => {
    if (instanceOpenedId && instanceId) {
      load(instanceId);
    }
  }, [instanceOpenedId, instanceId, triggerReload]);

  const data = tabSheetData(instanceId);

  const onLoadTabSheetRanges = (ranges?: PPSom.Rect[]): void => {
    loadTabSheetRanges(instanceId, ranges);
  };

  const onTabSheetPictureMouseDown = (args: ITabSheetPictureMouseDownArgs): void => {
    pressTabsheetPicture(instanceId, args);
  };

  return (
    <div className="table-visualizer" style={{ left: x, top: y, width, height }}>
      <Component
        className="table-visualizer-tabsheet"
        isFetching={isFetching}
        type={VisualizerType.Grid}
        browserRoutingMode={browserRoutingMode}
        editMode={ETabSheetEditMode.Disable}
        data={data}
        useNumberFormats
        isPtFontUnit
        zoom={zoom}
        width={width}
        height={height}
        rowHeaderWidth={0}
        columnHeaderHeight={1}
        onLoadRanges={onLoadTabSheetRanges}
        onPictureMouseDown={onTabSheetPictureMouseDown}
      />
    </div>
  );
};

Table.defaultProps = {
  triggerReload: undefined,
  width: undefined,
  height: undefined,
  zoom: undefined,
  browserRoutingMode: true,
  x: 0,
  y: 0,
};

export const mapStateToProps = (state: StatePart, ownProps: OwnProps): StateProps => ({
  instanceOpenedId: tableVisualizerSlice.selectors.openId(state, ownProps.instanceId),
  isFetching: tableVisualizerSlice.selectors.isFetching(state, ownProps.instanceId),
  tabSheetData: (instanceId: string): TabSheetRawData =>
    tableVisualizerSlice.selectors.tabSheetData(state, instanceId),
});

export const mapDispatchToProps: DispatchProps = {
  setOpenId: tableVisualizerSlice.actions.setOpenId,
  createInstance: tableVisualizerSlice.actions.createInstance,
  load: tableVisualizerThunk.load,
  pressTabsheetPicture: tableVisualizerThunk.pressTabsheetPicture,
  loadTabSheetRanges: tableVisualizerThunk.loadTabSheetRanges,
};

const enhance = compose(connect(mapStateToProps, mapDispatchToProps));
export default enhance(Table);
