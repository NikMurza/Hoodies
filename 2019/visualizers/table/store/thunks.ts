import { IAsyncHandler } from '@common/containers/spinner/store/thunks';
import bi from '@common/api/bi/index';
import { ITabSheetPictureMouseDownArgs } from '@common/components/lib/types/tab-sheet';
import { PPSom } from '@common/api/types/PPSom';
import { GetState } from './types';
import slice from './slice';

export const load = (instanceId: string): IAsyncHandler => async (
  dispatch: Function,
  getState: GetState,
): Promise<void> => {
  try {
    const state = getState();
    dispatch(slice.actions.setIsFetching({ instanceId, isFetching: true }));
    const openId = slice.selectors.openId(state, instanceId);

    const tabSheetData = await bi.tabSheet.getData(
      `${openId}!SemanticLayer!MetaModels!${instanceId}!Grid!TabSheet`,
    );

    dispatch(slice.actions.setTabSheetData({ instanceId, data: tabSheetData.rawData }));
  } finally {
    dispatch(slice.actions.setIsFetching({ instanceId, isFetching: false }));
  }
};

export const loadTabSheetRanges = (
  instanceId: string,
  ranges?: PPSom.TabRange[],
): IAsyncHandler => async (dispatch: Function, getState: GetState): Promise<void> => {
  const state = getState();
  const openId = slice.selectors.openId(state, instanceId);
  const tabSheetData = await bi.tabSheet.getRangesData(
    `${openId}!DataArea!Views!${instanceId}!TabSheet`,
    ranges,
  );
  dispatch(
    slice.actions.setTabSheetData({
      instanceId,
      data: tabSheetData,
    }),
  );
};

export const pressTabsheetPicture = (
  instanceId: string,
  args: ITabSheetPictureMouseDownArgs,
): IAsyncHandler => async (dispatch: Function, getState: GetState): Promise<void> => {
  const state = getState();
  const openId = slice.selectors.openId(state, instanceId);
  await bi.tabSheet.action.pressPictures(`${openId}!DataArea!Views!${instanceId}!TabSheet`, {
    coordinates: [args.coordinates],
  });
  dispatch(load(instanceId));
};

export default {
  load,
  pressTabsheetPicture,
  loadTabSheetRanges,
};
