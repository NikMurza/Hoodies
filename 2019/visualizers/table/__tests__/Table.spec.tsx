import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import '@testing-library/jest-dom';
import { act, render } from '@common/test-utils';
import { Table, mapStateToProps, mapDispatchToProps } from '../Table';
import { TabSheetRawData } from '../store/types';
import tableVisualizerSlice from '../store/slice';
import tableVisualizerThunk from '../store/thunks';

const initialState = {
  tableVisualizers: {
    testInstanceId: {
      openId: 'testOpenedId!M!S!OLFHKGABKBBAIFOAEENDAACPDHMKAIKJENLLACLKILCDEPEOG',
      tabSheetData: {},
      isFetching: false,
    },
  },
};

const props = {
  openedId: 'openedId',
  instanceKey: 'instanceKey',
  instanceId: '1',
  instanceOpenedId: 'instanceOpenedId',
  isFetching: false,
  tabSheetData: (): TabSheetRawData => ({}),
  setOpenId: jest.fn,
  createInstance: jest.fn,
  load: jest.fn,
  pressTabsheetPicture: jest.fn,
  loadTabSheetRanges: jest.fn,
};

describe('Table component', () => {
  it('render', async () => {
    await act(() => {
      render(<Table {...props} />);
    });
  });
  it('renders correctly', () => {
    expect(toJson(shallow(<Table {...props} />))).toMatchSnapshot();
  });

  const ownProps = {
    id: 'testId',
    instanceKey: 'testInstanceKey',
    openedId: 'testOpenedId',
    instanceId: 'testInstanceId',
    triggerReload: false,
    browserRoutingMode: false,
    x: 10,
    y: 20,
  };

  it('mapStateToProps', () => {
    expect(mapStateToProps(initialState, { ...ownProps }).instanceOpenedId).toEqual(
      tableVisualizerSlice.selectors.openId(initialState, ownProps.instanceId),
    );
    expect(mapStateToProps(initialState, { ...ownProps }).tabSheetData('testInstanceId')).toEqual(
      {},
    );
  });

  it('mapDispatchToProps', () => {
    expect(JSON.stringify(mapDispatchToProps.setOpenId)).toEqual(
      JSON.stringify(tableVisualizerSlice.actions.setOpenId),
    );
    expect(JSON.stringify(mapDispatchToProps.createInstance)).toEqual(
      JSON.stringify(tableVisualizerSlice.actions.createInstance),
    );

    expect(JSON.stringify(mapDispatchToProps.load)).toEqual(
      JSON.stringify(tableVisualizerThunk.load),
    );

    expect(JSON.stringify(mapDispatchToProps.pressTabsheetPicture)).toEqual(
      JSON.stringify(tableVisualizerThunk.pressTabsheetPicture),
    );

    expect(JSON.stringify(mapDispatchToProps.loadTabSheetRanges)).toEqual(
      JSON.stringify(tableVisualizerThunk.loadTabSheetRanges),
    );
  });
});
