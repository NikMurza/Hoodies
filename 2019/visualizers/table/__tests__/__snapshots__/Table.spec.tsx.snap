// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component renders correctly 1`] = `
<div
  className="table-visualizer"
  style={
    Object {
      "height": undefined,
      "left": 0,
      "top": 0,
      "width": undefined,
    }
  }
>
  <Component
    browserRoutingMode={true}
    className="table-visualizer-tabsheet"
    columnHeaderHeight={1}
    data={Object {}}
    editMode={0}
    isFetching={false}
    isPtFontUnit={true}
    onLoadRanges={[Function]}
    onPictureMouseDown={[Function]}
    rowHeaderWidth={0}
    type="Grid"
    useNumberFormats={true}
  />
</div>
`;
